### PR TITLE
feat: implement message filtering policy

### PR DESCRIPTION
**Pull request by guillaumelamirand**
---
**Gravitee.io Automatic Deployment**

🚀 A prerelease version of this package has been published on Gravitee's private artifactory, you can:
 - use it directly by updating your project with version: `1.0.0-apim-307-message-filtering-SNAPSHOT`
 - download it from Artifactory [here](https://odbxikk7vo-artifactory.services.clever-cloud.com/gravitee-snapshots/io/gravitee/policy/gravitee-policy-message-filtering/1.0.0-apim-307-message-filtering-SNAPSHOT/gravitee-policy-message-filtering-1.0.0-apim-307-message-filtering-SNAPSHOT.zip)

### CODE REVIEW

File: README.adoc [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+endif::[]
+
+== Phases
+
+[cols="4*", options="header"]
+|===
+^|onRequest
+^|onMessageRequest
+^|onResponseContent
+^|onMessageResponseContent
```
Comment (ytvnr) [inline README.adoc]:
```suggestion
^|onMessageResponse
```

File: src/main/java/io/gravitee/policy/messagefiltering/MessageFilteringPolicy.java [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    @Override
+    public String id() {
+        return "message-filtering";
+    }
+
+    @Override
+    public Completable onMessageRequest(MessageExecutionContext ctx) {
+        return Completable.defer(
+            () -> {
+                String computedFilter = computeFilter(ctx);
```
Comment (ytvnr) [inline src/main/java/io/gravitee/policy/messagefiltering/MessageFilteringPolicy.java:44]: I would have put the call to computeFilter() inside filter() method as it's the first required step for the filter to compute what we have in the configuration correctly

but it's not a blocker

File: src/main/java/io/gravitee/policy/messagefiltering/MessageFilteringPolicy.java [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+                return ctx.request().onMessage(message -> filter(ctx, computedFilter, message));
+            }
+        );
+    }
+
+    @Override
+    public Completable onMessageResponse(MessageExecutionContext ctx) {
+        return Completable.defer(
+            () -> {
+                String computedFilter = computeFilter(ctx);
```
Comment (ytvnr) [inline src/main/java/io/gravitee/policy/messagefiltering/MessageFilteringPolicy.java:54]: same

**Review by ytvnr — COMMENTED** [review]
a change in the readme and a suggestion in the code

File: src/main/java/io/gravitee/policy/messagefiltering/MessageFilteringPolicy.java (hunk 2 above)
Comment (guillaumelamirand) [reply src/main/java/io/gravitee/policy/messagefiltering/MessageFilteringPolicy.java:44]: I don"t understand you comment. computeFilter is not about filtering the ctx but is about computing the EL filter expression. Does it make sense?

File: src/main/java/io/gravitee/policy/messagefiltering/MessageFilteringPolicy.java (hunk 3 above)
Comment (guillaumelamirand) [reply src/main/java/io/gravitee/policy/messagefiltering/MessageFilteringPolicy.java:54]: same as above.

File: src/main/java/io/gravitee/policy/messagefiltering/MessageFilteringPolicy.java (hunk 2 above)
Comment (guillaumelamirand) [reply src/main/java/io/gravitee/policy/messagefiltering/MessageFilteringPolicy.java:44]: OK after re-read I got it. I don't want to compute/check the filter on each message, only once for all. That why the computation is outside the filter method.

Comment (ytvnr) [reply src/main/java/io/gravitee/policy/messagefiltering/MessageFilteringPolicy.java:44]: yeah indeed, I understand better.

**Pull request merged**